File: leetcode/go-lc/049_group_anagrams.go
package main

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

func groupAnagrams(strs []string) [][]string {
	hm := make(map[string][]string)

	for _, str := range strs {
		l_str := strings.Split(str, "")
		sort.Strings(l_str)
		s_str := strings.Join(l_str, "")
		hm[s_str] = append(hm[s_str], str)
	}

	res := make([][]string, 0)
	for _, v := range hm {
		res = append(res, v)
	}

	return res
}

func main() {
	// Convert the test inputs to []interface{} format
	input := make([][]interface{}, 3)

	input[0] = []interface{}{[]string{"eat", "tea", "tan", "ate", "nat", "bat"}}
	input[1] = []interface{}{[]string{""}}
	input[2] = []interface{}{[]string{"a"}}

	expected := []interface{}{[][]string{{"eat", "tea", "ate"}, {"tan", "nat"}, {"bat"}}, [][]string{{""}}, [][]string{{"a"}}}

	start := time.Now()

	for j := 0; j < len(input); j++ {
		result := groupAnagrams(input[j][0].([]string))

		for i := 0; i < len(result); i++ {
			if len(result[i]) != len(expected[j].([][]string)[i]) {
				fmt.Println("Test case ", j, " failed")
				return
			}
		}
	}

	elapsed := time.Since(start).Seconds()
	fmt.Println("049. groupAnagrams", elapsed)
}

File: leetcode/python-lc/049_group_anagrams.py
from typing import Dict, List
from base import TestCase


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hm: Dict[str, List[str]] = {}

        for s in strs:
            s_sorted = "".join(sorted(s))
            if s_sorted not in hm:
                hm[s_sorted] = []
            hm[s_sorted].append(s)
        return list(hm.values())


if __name__ == "__main__":
    solution = Solution()

    testcase = TestCase(
        "049",
        "groupAnagrams",
        solution,
        [(["eat", "tea", "tan", "ate", "nat", "bat"],)],
        [[["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]],
    )
    testcase.timeit()
